Fixes parameter after fetch

Vignesh1-art opened the pull request:
Thank you for submitting this pull request! We really appreciate you spending the time to work on these changes.

## What is the motivation?

Parsing fails if variables are encountered after `FETCH` statement

## What does this change do?

Fixes this problem. I have added new parsing function to parse list of variables and Parses it after `FETCH` statement and implemented to compute variable

## What is your testing strategy?

Added unit tests

## Is this related to any issues?

https://github.com/surrealdb/surrealdb/issues/3028

<!-- Use 'Closes' or 'Fixes' to mark that this pull request successfully closes an issue. -->

## Does this change need documentation?

If this pull request requires changes, updates, or improvements to the documentation, then add a corresponding issue on the [docs.surrealdb.com](https://github.com/surrealdb/docs.surrealdb.com) repository, and link to it here.

<!-- Delete one of the following lines as necessary, and enter the correct corresponding issue number. -->

- [x] No documentation needed

## Have you read the Contributing Guidelines?

<!-- All pull requests require that the contributing guidelines have been read and agreed to. -->

- [x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

Vignesh1-art commented:
@tobiemh  This PR has required test case can you please review it.

Vignesh1-art commented:
> @tobiemh This PR has required test case. Can you please review it.

@tobiemh

Vignesh1-art commented:
@tobiemh or @DelSkayn someone Please review and approve the workflow

Vignesh1-art commented:
I have fixed formatting issues, @tobiemh  or @DelSkayn  can you please approve the workflow

Vignesh1-art commented:
Also @tobiemh or @DelSkayn can you please tell me how to  check for formatting issues in local?

gguillemas commented:
Hi @Vignesh1-art. We really appreciate your contribution. Please, be patient and do not repeatedly ping people, they have personal lives and can be sick, on holidays or just busy with something else. You can check format locally using `cargo fmt --all --check`, which will also automatically fix the format if you remove `--check`.

Vignesh1-art commented:
> Hi @Vignesh1-art. We really appreciate your contribution. Please, be patient and do not repeatedly ping people, they have personal lives and can be sick, on holidays or just busy with something else. You can check format locally using `cargo fmt --all --check`, which will also automatically fix the format if you remove `--check`.

I will take care of it from next time! Thanks @gguillemas, I ran formatting checks in local and now no issues found.

Vignesh1-art commented:
I rebased the branch and did not try to compile it mine bad sorry Now It is working

Vignesh1-art commented on core/src/sql/value/fetch.rs:
```diff
@@ -35,6 +35,16 @@ impl Value {
 						Some(v) => stk.run(|stk| v.fetch(stk, ctx, opt, path.next())).await,
 						None => Ok(()),
 					},
+					Part::Value(f) => {
+						let x = stk.run(|stk| f.compute(stk, ctx, opt, None)).await?;
+						match x {
+							Value::Strand(s) => match v.get_mut(s.as_str()) {
+								Some(v) => stk.run(|stk| v.fetch(stk, ctx, opt, path)).await,
+								_ => Ok(()),
+							},
+							_ => Ok(()),
+						}
+					}
```
I can do the change

Vignesh1-art commented:
@tobiemh  Can you suggest please?

tobiemh commented:
Hi @Vignesh1-art changes would need to be made according to this review here: https://github.com/surrealdb/surrealdb/pull/4086#discussion_r1618816931

Vignesh1-art commented:
@tobiemh Yes i have made changes according that comment, but since I have changed Fetch(Idiom) to Fetch(Value) I can't pass it to compute of fetch statement, because it is of not type Part, mine question is should write logic to convert Value to Part? So logic is

if Value is idiom then
converting to part is straight forward 
else 
compute Value this will return Strand, Number or Something
Convert that to Part.

Is that correct @tobiemh ?

DelSkayn commented on core/src/syn/parser/idiom.rs:
```diff
@@ -239,6 +239,7 @@ impl Parser<'_> {
 				let graph = ctx.run(|ctx| self.parse_graph(ctx, Dir::In)).await?;
 				Part::Graph(graph)
 			}
+			t!("$param") => Part::Value(Value::Param(self.next_token_value()?)),
```
Adding support for a parameter here doesn't just change the `FETCH` clause to support params but changes every place where a idiom is allowed which is a lot of places, most of which I am not sure we want to support having parameters in those positions.

Instead of adding support here for parameters I think a better approach would be to implement it in `core::syn::parser::stmt::parts::try_parse_fetch`. That way the change is limited to just the `FETCH` clause.

Vignesh1-art commented:
Hi @DelSkayn ,
Thanks for taking time to review my PR. In parser I have done the changes as you suggested. I need your help in compute part of it. As @tobiemh suggested here https://github.com/surrealdb/surrealdb/pull/4086#discussion_r1618816931 I have changed the Fetch struct to take Value.

Now here

	async fn output_fetch(
		&mut self,
		stk: &mut Stk,
		ctx: &Context<'_>,
		opt: &Options,
		stm: &Statement<'_>,
	) -> Result<(), Error> {
		if let Some(fetchs) = stm.fetch() {
			for fetch in fetchs.iter() {
				let value = fetch.compute(stk, ctx, opt, None).await?;
				let mut values = self.results.take()?;
				// Loop over each result value
				for obj in &mut values {
					// Fetch the value at the path
					stk.run(|stk| obj.fetch(stk, ctx, opt, value)).await?;
				}
				self.results = values.into();
			}
		}
		Ok(())
	}

I have computed the fetch clause here (Hopefully I have done it in right place :) ). But now problem is obj.fetch won't accept fetch as parameter as type required for it is different now since I have changed Fetch to pub struct Fetch(pub Value); I need your suggestion here how to proceed further :). Do I need to convert Value to Part ?
And also now I released I should conditionally compute fetch, I mean if it is Param only then compute. Is it correct?

Please checkout this comment also https://github.com/surrealdb/surrealdb/pull/4086#issuecomment-2149053532

Vignesh1-art commented:
@DelSkayn  I have pushed the changes can you review now

DelSkayn commented on core/src/sql/statements/output.rs:
```diff
@@ -39,7 +40,23 @@ impl OutputStatement {
+						let p: Part = Part::Field(Ident(s.0));
+						new_idiom = Idiom(vec![p]);
+						i = &new_idiom;
+					} else {
+						break;
+					}
+				} else {
+					break;
+				}
```
Breaking here is probably not the right behavior, this will cause an invalid production to just ignore the fetch without giving any indication back to the user.

If the param value is not a string you should probably return an error that it must be a string, and if the value is neither an param, nor an idiom, then it the query is invalid so you should probably return the Error::Unreachable.

DelSkayn commented on core/src/syn/parser/test/streaming.rs:
```diff
@@ -590,9 +590,9 @@ fn statements() -> Vec<Statement> {
-				Ident("RETURN".to_owned()).to_owned(),
-			)]))])),
+			fetch: Some(Fetchs(vec![Fetch(Value::Idiom(Idiom(vec![Part::Field(
+				Ident("RETRUN".to_owned()).to_owned(),
```
Why did you change this?

DelSkayn commented on core/src/syn/parser/stmt/parts.rs:
```diff
@@ -107,8 +107,11 @@ impl Parser<'_> {
 		}
-		let v = self.parse_idiom_list(ctx).await?.into_iter().map(Fetch).collect();
-		Ok(Some(Fetchs(v)))
+		let mut fetchs = vec![Fetch(self.parse_value_field(ctx).await?)];
+		while self.eat(t!(",")) {
+			fetchs.push(Fetch(self.parse_value_field(ctx).await?));
+		}
+		Ok(Some(Fetchs(fetchs)))
```
This will parse all possible values after fetch, including subqueries, blocks statements etc. This should be more restricted, testing for a parameter only, otherwise parsing an idiom.

Vignesh1-art commented:
@DelSkayn  I have done changes as you mentioned

DelSkayn reviewed (dismissed):
Alright! Everything looks good now.

Vignesh1-art commented:
> Alright! Everything looks good now.

I have done rebased branch with main to fix conflict so please approve again. Thank you for your time.

Vignesh1-art commented:
@DelSkayn  Thank you for making changes. All checks are good.

Vignesh1-art commented:
@DelSkayn  Thank you for making changes, all checks passed it looks good now.

emmanuel-keller reviewed (approved):
LGTM!
